packages/astro/src/core/build/generate.ts

```diff
@@ -175,65 +164,61 @@ async function generatePath(pathname: string, opts: StaticBuildOptions, gopts: G
-					// Return this as placeholder, which will be ignored by the browser.
-					// TODO: In the future, we hope to run this entire script through Vite,
-					// removing the need to maintain our own custom Vite-mimic resolve logic.
-					if (specifier === BEFORE_HYDRATION_SCRIPT_ID) {
-						return 'data:text/javascript;charset=utf-8,//[no before-hydration script]';
-					}
-					throw new Error(`Cannot find the built path for ${specifier}`);
+	const options: RenderOptions = {
+		legacyBuild: false,
+		links,
+		logging,
+		markdownRender: astroConfig.markdownOptions.render,
+		mod,
+		origin,
+		pathname,
+		scripts,
+		renderers,
+		async resolve(specifier: string) {
+			const hashedFilePath = internals.entrySpecifierToBundleMap.get(specifier);
+			if (typeof hashedFilePath !== 'string') {
+				// If no "astro:scripts/before-hydration.js" script exists in the build,
+				// then we can assume that no before-hydration scripts are needed.
+				// Return this as placeholder, which will be ignored by the browser.
+				// TODO: In the future, we hope to run this entire script through Vite,
+				// removing the need to maintain our own custom Vite-mimic resolve logic.
+				if (specifier === BEFORE_HYDRATION_SCRIPT_ID) {
+					return 'data:text/javascript;charset=utf-8,//[no before-hydration script]';
 				}
-				const relPath = npath.posix.relative(pathname, '/' + hashedFilePath);
-				const fullyRelativePath = relPath[0] === '.' ? relPath : './' + relPath;
-				return fullyRelativePath;
-			},
-			method: 'GET',
-			headers: new Headers(),
-			route: pageData.route,
-			routeCache,
-			site: astroConfig.buildOptions.site,
-			ssr: opts.astroConfig.buildOptions.experimentalSsr,
-		};
-
-		let body: string;
-		if (pageData.route.type === 'endpoint') {
-			const result = await callEndpoint(mod as unknown as EndpointHandler, options);
-
-			if (result.type === 'response') {
-				throw new Error(`Returning a Response from an endpoint is not supported in SSG mode.`);
+				throw new Error(`Cannot find the built path for ${specifier}`);
 			}
-			body = result.body;
-		} else {
-			const result = await render(options);
+			const relPath = npath.posix.relative(pathname, '/' + hashedFilePath);
+			const fullyRelativePath = relPath[0] === '.' ? relPath : './' + relPath;
+			return fullyRelativePath;
+		},
+		method: 'GET',
+		headers: new Headers(),
+		route: pageData.route,
+		routeCache,
+		site: astroConfig.buildOptions.site,
+		ssr: opts.astroConfig.buildOptions.experimentalSsr,
+	};
 
-			// If there's a redirect or something, just do nothing.
-			if (result.type !== 'html') {
-				return;
-			}
-			body = result.html;
+	let body: string;
+	if (pageData.route.type === 'endpoint') {
+		const result = await callEndpoint(mod as unknown as EndpointHandler, options);
+
+		if (result.type === 'response') {
+			throw new Error(`Returning a Response from an endpoint is not supported in SSG mode.`);
```

Comment:
Oh good catch!

Reply:
I wish I came up with this! but it was just moved from elsewhere. You probably have @matthewp to thank :)